### PR TITLE
CSPL-793: Add documentation for App Framework

### PR DESCRIPTION
**Pull request by sgontla**

### CODE REVIEW

File: docs/Examples.md [hunk 1]
```diff
@@ -413,6 +413,11 @@ including the `apps_location` parameter in your default settings. The value
 may either be a comma-separated list of apps or a YAML list, with each app
```
Comment (eddikay) [inline docs/Examples.md:413]: For line 410, can we insert a statement in support of the new App Framework and suggest its use?

With the latest release of the Splunk Operator, an App Framework is available to centrally store and deploy apps. See AppFramework.md for information and examples.

File: docs/Examples.md [hunk 2]
```diff
@@ -413,6 +413,11 @@ including the `apps_location` parameter in your default settings. The value
...
 referenced using a filesystem path or URL.
 
+Note: In the case of `SearchHeadCluster` or `ClusterMaster` when the apps are configured through 
+the `apps_location`, all those apps will be deployed to the Search Heads or Indexers respectively.
+In order to install the apps locally to the Deployer or ClusterMaster, the apps should be specified through
```
Comment (eddikay) [inline docs/Examples.md:418]: To install the apps locally to the Deployer or ClusterMaster, the apps should be specified through `apps_location_local`.

File: docs/Examples.md [hunk 3]
```diff
@@ -445,6 +452,33 @@ spec:
         - "/mnt/apps/app1.tgz"
         - "/mnt/apps/app2.tgz"
 ```
+In the above example, app1 and app2 are installed on Splunk Standalone instances.
```
Comment (eddikay) [inline docs/Examples.md:455]: Move this to 436:

In the standalone example, app1 and app2 are installed on Splunk Standalone instances.

File: docs/Examples.md [hunk 4]
```diff
@@ -445,6 +452,33 @@ spec:
...
+    splunk:
+      apps_location:
+        - "/mnt/apps/app3.tgz"
+        - "/mnt/apps/app4.tgz"
+      apps_location_local:
+        - "/mnt/apps/app5.tgz"
+        - "/mnt/apps/app6.tgz"
+```
+
+In the above example, app3 and app4 are installed on Splunk Indexer instances of the Indexer Cluster for which it is Cluster Master. app5 and app6 are installed locally on the ClusterMaster instance.
```
Comment (eddikay) [inline docs/Examples.md:481]: Move to 458:
In the ClusterMaster example, app3 and app4 are installed on any indexer instances that are managed by the cluster master. App5 and app6 are installed locally on the ClusterMaster instance.

File: docs/Examples.md [hunk 5]
```diff
@@ -408,11 +408,17 @@ for more details.
...
+With the latest release of the Splunk Operator, an App Framework is available to centrally store and deploy apps. See AppFramework.md for information and examples.
+
 The Splunk Operator can be used to automatically install apps for you by
 including the `apps_location` parameter in your default settings. The value
 may either be a comma-separated list of apps or a YAML list, with each app
 referenced using a filesystem path or URL.
 
+Note: In the case of `SearchHeadCluster` or `ClusterMaster` when the apps are configured through 
+the `apps_location`, all those apps will be deployed to the Search Heads or Indexers respectively.
+To install the apps locally to the Deployer or ClusterMaster, the apps should be specified through apps_location_local.
```
Comment (jryb) [inline docs/Examples.md:420]: nit: `apps_location_local`

File: docs/Examples.md [hunk 6]
```diff
@@ -427,6 +433,9 @@ kubectl create configmap splunk-apps --from-file=app1.tgz --from-file=app2.tgz
 You can have the Splunk Operator install these automatically using something
 like the following:
 
+Example: Standalone
```
Comment (jryb) [inline docs/Examples.md:436]: Should we put this in a heading?

File: docs/AppFramework.md [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+6. Apply the Customer Resource specification: `kubectl -f apply ClusterMaster.yaml`
+
+
+### How to use the App Framework on Search Head Cluster
+
+This example, describes installation of apps on Search Head Cluster as well as Deployer. This is achieved by deploying a SearchHeadCluster CR with a storage volume, the location of the app archives, and set the installation scope to support both local and cluster app distribution.
+
+1. Confirm your S3-based storage volume path and URL.
+2. Create a Kubernetes Secret Object with the storage credentials. 
+   * Example: `kubectl create secret generic s3-secret --from-literal=s3_access_key=$7$FLa0GfJNEzdcqLBcZmW1SDg==`
```
Comment (jryb) [inline docs/AppFramework.md:127]: We verified that we only need the access key, correct?

Comment (sgontla) [reply docs/AppFramework.md:127]: Seems like I jumped on this early. Our code fails without this. May be we need to give a try, and understand it more.

**Review by jryb — APPROVED** [review]

**Comment by sgontla** [comment]
> Not sure if my comment was noticed. Please move 173 - 177 to line 119.

Thanks for pointing it. Addressed it for both SHC and IDXC cases.

File: docs/AppFramework.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+`appsRepoPollIntervalSeconds` helps configure the polling interval(in seconds) to detect addition or modification of apps on the Remote Storage
+
+
+
+## App Framework Limitations
+
+The App Framework does not review, preview, analyze, or enable Splunk Apps and Add-ons. The administrator is responsible for previewing the app or add-on contents, verifying the app is enabled, and that the app is supported with the version of Splunk Enterprise used in the containers. For App packaging specifications see [Package apps for Splunk Cloud or Splunk Enterprise](https://dev.splunk.com/enterprise/docs/releaseapps/packageapps/) in the Splunk Enterprise Developer documentation. The app archive files must end with .spl or .tgz; all other files are ignored.
+
+1. The App Framework has no support to remove an app or add-on once it’s been deployed. To disable an app, update the archive contents located in the App Source, and set the app.conf state to disabled. If the container instance is running Splunk Web, you can also use the Apps UI to disable an app on an individual container.
```
Comment (smohan-splunk) [inline docs/AppFramework.md:316]: Can we please refrain from recommending UI for making changes to App?

File: docs/Examples.md [hunk 9]
```diff
@@ -408,11 +408,17 @@ for more details.
 
 *Note that this requires using the Splunk Enterprise container version 8.1.0 or later*
 
+With the latest release of the Splunk Operator, an App Framework is available to centrally store and deploy apps. See [AppFramework](AppFramework.md) for information and examples.
```
Comment (smohan-splunk) [inline docs/Examples.md:411]: Can we say "a Beta version of the new App framework"

Also, add something like: 
The below method of installing apps continues to be supported, but will be deprecated in future releases.

**Review by smohan-splunk — APPROVED** [review]

File: docs/AppFramework.md (hunk 8 above)
Comment (sgontla) [reply docs/AppFramework.md]: My bad, we already removed it, but missed it when i am combining them.

File: docs/AppFramework.md [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    | IndexerCluster    | N/A            | No                    |
+
+* `volume` refers to the remote storage volume name configured under the `volumes` stanza (see previous section)
+* `location` helps configure the specific appSource present under the `path` within the `volume`, containing the apps to be installed  
+
+### appsRepoPollIntervalSeconds
+
+`appsRepoPollIntervalSeconds` helps configure the polling interval(in seconds) to detect addition or modification of apps on the Remote Storage
+
+
```
Comment (smohan-splunk) [inline docs/AppFramework.md:310]: Add impact on livenessInitialDelaySeconds & readinessInitialDelaySeconds

**Review by smohan-splunk — APPROVED** [review]

**Review by eddikay — APPROVED** [review]

File: docs/AppFramework.md [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    volumes:
+      - name: volume_app_repo_us
+        storageType: s3
+        provider: aws
+        path: bucket-app-framework-us-west-2/shcLoc-us/
+        endpoint: https://s3-us-west-2.amazonaws.com
+        secretRef: s3-secret
+```
+
+6. Apply the Customer Resource specification: `kubectl -f apply SearchHeadCluster.yaml`
```
Comment (gaurav-splunk) [inline docs/AppFramework.md:176]: ditto

File: docs/AppFramework.md [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+### appRepo
+
+`appRepo` is the start of the App Framework specification, and contains all the configurations required for App framework to be successfully configured
+
+### volumes
+
+`volumes` helps configure the remote storage volumes. App framework expects apps that are to be installed in various Splunk deployments to be hosted in one or more remote storage volumes
+
+* `name` uniquely identifies the volume name within a CR. This is to locally used by the Operator to identify the volume
+* `storageType` describes the type of remote storage. Currently `s3` i:w
```
Comment (gaurav-splunk) [inline docs/AppFramework.md:278]: nit: i:w ??

File: docs/AppFramework.md [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+* `volume` refers to the remote storage volume name configured under the `volumes` stanza (see previous section)
+* `location` helps configure the specific appSource present under the `path` within the `volume`, containing the apps to be installed  
+
+### appsRepoPollIntervalSeconds
+
+`appsRepoPollIntervalSeconds` helps configure the polling interval(in seconds) to detect addition or modification of apps on the Remote Storage
+
+## Impact of livenessInitialDelaySeconds and readinessInitialDelaySeconds
+
+* Splunk Operator CRDs support the configuration of initialDelaySeconds(insert link to https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) for both Liveliness (livenessInitialDelaySeconds) and Readiness (readinessInitialDelaySeconds) probes
```
Comment (gaurav-splunk) [inline docs/AppFramework.md:312]: should we remove this - `insert link to`  and actually add a link?

File: docs/AppFramework.md [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+## App Framework Limitations
+
+The App Framework does not review, preview, analyze, or enable Splunk Apps and Add-ons. The administrator is responsible for previewing the app or add-on contents, verifying the app is enabled, and that the app is supported with the version of Splunk Enterprise used in the containers. For App packaging specifications see [Package apps for Splunk Cloud or Splunk Enterprise](https://dev.splunk.com/enterprise/docs/releaseapps/packageapps/) in the Splunk Enterprise Developer documentation. The app archive files must end with .spl or .tgz; all other files are ignored.
+
+1. The App Framework has no support to remove an app or add-on once it’s been deployed. To disable an app, update the archive contents located in the App Source, and set the app.conf state to disabled.
+
+2. The App Framework tracks the app installation state per CR. If you configure a Standalone CR to use more than one replicas, a new replica Pod will not receive any apps that were previously deployed.
```
Comment (gaurav-splunk) [inline docs/AppFramework.md:322]: We need to change this now.

**Review by gaurav-splunk — CHANGES_REQUESTED** [review]

**Review by jryb — APPROVED** [review]

**Review by gaurav-splunk — APPROVED** [review]

**Pull request merged**